### ISSUE TITLE
Consider migrating from pcre to pcre2

### ISSUE THREAD

**Issue by alex65536** [nim-lang/Nim#23668]
### Description

Nim depends on the obsolete PCRE library, which is no longer supported with last release in 2021, and should migrate to PCRE2 library available since 2015.

I am concerned that Debian wants to remove the deprecated PCRE library (see [this](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1071970) bug report), making some of the functionality and some of the Nim packages unusable in the next stable Debian release.

So, please consider porting `std/nre` and `std/re` to PCRE2. PCRE and PCRE2 seem to be API incompatible, but it should not be a problem for anyone who doesn't use the C wrappers directly.

### Nim Version

Any version at the time of writing (`devel` branch is affected as of `cdfc886f88cc95a8dc05e2211d1030b146a521f5`)

**Comment by ringabout** [comment]
side note: ubuntu 24.04 doesn't seem to ship `libpcre3` by default, you have to install `pcre` manually => https://github.com/nim-lang/Nim/pull/23649

**Comment by arnetheduck** [comment]
This is an excellent opportunity to deprecate the re module and recommend a library instead, such as https://github.com/nitely/nim-regex (this way the library can be updated without waiting months for a release and getting hit by other unrelated changes)

**Comment by alex65536** [comment]
> This is an excellent opportunity to deprecate the re module and recommend a library instead

Well, part of Nim depends on regexes itself. I took a quick look over the code base and discovered two such places in Nim itself: `nimgrep` and tests for `nimsuggest`. The latter one may be easy to remove, but for `nimgrep` using regexes is the core functionality, and the Nim repo seems to avoid dependencies on Nimble packages for now.

**Comment by arnetheduck** [comment]
> Nim repo seems to avoid dependencies on Nimble packages for now.

actually it does not - some packages are depended upon and downloaded to a special `dist` folder to create a "full" source distribution - the same can be done for regex, to create stand-alone compiler code releases.

**Comment by jmgomez** [comment]
+1 to deprecate it and to use an external package. I don't mind to have it in the std. The issue I see is the lib dep

**Comment by arnetheduck** [comment]
also, `nimgrep` could easily be a package - there's literally nothing in there that needs to live in nim proper.

**Comment by nileshpatra** [comment]
Did you get a chance to work on it? Nim has been out of debian testing for a while by now due to this being a release critical bug for us (pcre3) not being a part of next release.

**Comment by Araq** [comment]
FWIW I found no migration guide.

**Comment by nitely** [comment]
fwiw, there is this https://github.com/PCRE2Project/pcre2/issues/51#issuecomment-973011802 and the linked PRs as migration examples.

**Comment by SuperSandro2000** [comment]
NixOS is also phasing out pcre https://github.com/NixOS/nixpkgs/issues/356387

**Comment by andreasstieger** [comment]
So is openSUSE: https://lists.opensuse.org/archives/list/factory@lists.opensuse.org/thread/BK3SPPFOM3LI6K5PTXPKZMKMIUIPOEXS/

**Comment by inv2004** [comment]
opensuse tumbleweed does not have `libpcre.so.(.1|.3)` anymore

**Comment by andreasstieger** [comment]
Correct. OpenSUSE dropped PCRE, and carries the patch from #24405 for nim.

**Comment by kostya** [comment]
ubuntu 26.04 also remove libpcre3-dev

could not load: libpcre.so(.3|.1|)
(compile with -d:nimDebugDlOpen for more information)
